### DISCUSSION TITLE
can i popup a menu to center of screen

### DISCUSSION THREAD

**Discussion by ffs-akl** [LBCrion/sfwbar#272] [Q&A] [answered]
I would like to make a power menu.   i  hope it can popup a menu to center of screen, when i click on it's icon.
Now the hard part is that I don't know how to control where it pops up

**Comment by LBCrion** [comment]
The proper answer in wayland is "no". An application can't control placement of it's windows. I may be possible to use some heuristics for a layer shell application to control location of it's xdg popup, but it would be hacky and won't work for all configurations (with potentially very messy results when it doesn't work).

**Comment by Consolatis** [comment, marked as answer]
The question is also what `menu` refers to. If you mean something like a context menu then no as explained above. If you mean a bigger surface with buttons and the like that might be possible. See this python snippet:
```py
#!/usr/bin/env python3

import gi
gi.require_version("Gtk", "3.0")
gi.require_version('GtkLayerShell', '0.1')
from gi.repository import Gtk, GtkLayerShell

if __name__ == '__main__':
	win = Gtk.Window()
	GtkLayerShell.init_for_window(win)
	GtkLayerShell.set_layer(win, GtkLayerShell.Layer.TOP)
	win.set_size_request(640, 480)
	win.show()
	Gtk.main()
```

This uses the fact that the wlr-layershell protocol states that layershell surfaces [without anchors are centered](https://gitlab.freedesktop.org/wlroots/wlr-protocols/-/blob/master/unstable/wlr-layer-shell-unstable-v1.xml#L139).

**Reply by ffs-akl** [reply]
I mean a bigger surface with buttons like the dialog of lxqt-leave.  thank you!

**Comment by LBCrion** [comment]
Gtk menu is an xdg popup which is positioned relative to the patent
window.  An unanchored layer shell window typically is placed in the center
of the screen,  but there is no guarantee it will be. The placement is up
to a compositor.

**Reply by Consolatis** [reply]
I'd say if a compositor places a layershell surface without anchors anywhere but in the center its a bug in the compositor. At least from how I understand the protocol.

**Comment by LBCrion** [comment]
Indeed,  you're right. From the protocol: "... or in the center if none is
specified." (In respect to the anchors).

**Comment by LBCrion** [comment]
That shouldn't be too difficult, The anchors are currently set via bar size
and bar direction settings. In theory we could add a setting to unanchor a
bar. (this is how the switcher works) Then you could create a bar in the
middle of the screen and populate then with widgets of your choice.

**Reply by ffs-akl** [reply]
Sounds great. Maybe we should dig deeper.

**Comment by sfs-pra** [comment]
https://github.com/loserMcloser/waylogout ?

**Reply by ffs-akl** [reply]
Great for  my scenarios.

**Comment by sfs-pra** [comment]
Labwc. The https://github.com/v1cont/yad or https://gitlab.gnome.org/GNOME/zenity box appears in the middle of the screen. Is it possible to move it to the mouse cursor or to the bottom right corner. In Xorg you could do the same with https://web.archive.org/web/20221116231622/http://tripie.sweb.cz/utils/wmctrl/

**Reply by 01micko** [reply]
In `labwc` you can use a `windowRule` in `rc.xml

```xml
   <windowRule identifier="app_id*" matchOnce="false">
      <action name="AutoPlace" policy="cursor"/>
    </windowRule>
```

https://github.com/labwc/labwc/blob/master/docs/labwc-config.5.scd#placement

**Comment by sfs-pra** [comment]
Added to the beginning of rc.xml :
```
<windowRule identifier="yad" matchOnce="false">
      <action name="AutoPlace" policy="cursor"/>
</windowRule>
```
yad in the center of the screen

**Reply by 01micko** [reply]
Try this `yad --name=whatever --calendar` and replace 'yad' with whatever in the rc.xml. The name sets the app_id for wayland. (and icon in sfwbar)

**Comment by sfs-pra** [comment]
`<windowRule identifier="whatever*" matchOnce="false">` it didn't work
